blockchain/chain.go

```diff
@@ -1143,94 +1170,150 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
-	// violated.
-	fork := b.bestChain.FindFork(targetTip)
-	reorgErr := b.reorganizeChainInternal(targetTip)
-	if reorgErr != nil {
-		if err := b.reorganizeChainInternal(origTip); err != nil {
-			panicf("failed to reorganize back to known good chain tip %s "+
-				"(height %d): %v -- probable database corruption", origTip.hash,
-				origTip.height, err)
+		// Determine if the chain is being reorganized to a competing branch.
+		// This is the case when the current tip is not an ancestor of the
+		// target tip.
+		if !sentReorgingNtfn && target.Ancestor(tip.height) != tip {
+			// Send a notification announcing the start of the chain
+			// reorganization.
+			//
+			// Notice that the chain lock is not released before sending the
+			// notification.  This is intentional and must not be changed
+			// without understanding why!
+			b.sendNotification(NTChainReorgStarted, nil)
+			sentReorgingNtfn = true
+
+			defer func() {
+				// Send a notification announcing the end of the chain
+				// reorganization.
+				//
+				// Notice that the chain lock is not released before sending the
+				// notification.  This is intentional and must not be changed
+				// without understanding why!
+				b.sendNotification(NTChainReorgDone, nil)
+			}()
 		}
 
-		return reorgErr
+		// Attempt to reorganize the chain to the new tip.  In the case it
+		// fails, attempt to reorganize to the best valid block with the most
+		// cumulative proof of work instead.
+		err := b.reorganizeChainInternal(target)
+		if err != nil {
+			// Shutting down.
+			if errors.Is(err, errInterruptRequested) {
+				return err
+			}
+
+			// Typically, if a reorganize fails, there will only be a single
+			// error due to the block that caused the failure.  However, it is
+			// possible that several candidate branches might fail.  Thus, track
+			// them all so they can potentially be converted to a multi error
+			// later if needed.
+			reorgErrs = append(reorgErrs, err)
+
+			// Determine a new best candidate since the reorg failed.  This
+			// should realistically always result in a different target than the
+			// current one unless there is some type of unrecoverable error,
+			// such as a disk failure.  In that case, bail out to avoid
+			// attempting to do the same reorg over and over.
+			newTarget := b.index.FindBestChainCandidate()
+			if newTarget == target {
+				break
+			}
+			target = newTarget
+		}
 	}
 
-	// Send a notification that a blockchain reorganization took place.
-	reorgData := &ReorganizationNtfnsData{origTip.hash, origTip.height,
-		targetTip.hash, targetTip.height}
-	b.chainLock.Unlock()
-	b.sendNotification(NTReorganization, reorgData)
-	b.chainLock.Lock()
+	// Potentially update whether or not the chain believes it is current based
+	// on the new tip.  Notice that the tip is reset to whatever the best chain
+	// actually is here versus using the one from above since it might not match
+	// reality if there were errors during reorg.
```

Comment:
Went over this code a few times, and thanks again for implementing these changes. Only noticed a single comment that read strangely and not worth holding up the PR over.

```suggestion
	// reality if there were errors during the reorg.
```

Reply:
There might have been more than one, so it can't really be "the" reorg.  I think it's correct as it is too, but I went ahead and changed it to "while reorganizing" to avoid any ambiguity.